### PR TITLE
[SPARK-43071][SQL] Support SELECT DEFAULT with ORDER BY, LIMIT, OFFSET for INSERT source relation

### PR DESCRIPTION
**Pull request by dtenedor**
### What changes were proposed in this pull request?

This PR extends column default support to allow the ORDER BY, LIMIT, and OFFSET clauses at the end of a SELECT query in the INSERT source relation.

For example:

```
create table t1(i boolean, s bigint default 42) using parquet;
insert into t1 values (true, 41), (false, default);
create table t2(i boolean default true, s bigint default 42, 
                t string default 'abc') using parquet;
insert into t2 (i, s) select default, s from t1 order by s limit 1;
select * from t2;
> true, 41L, "abc"
```

### Why are the changes needed?

This improves usability and helps prevent confusing error messages.

### Does this PR introduce _any_ user-facing change?

Yes, SQL queries that previously failed will now succeed.

### How was this patch tested?

This PR adds new unit test coverage.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala [hunk 1]
```diff
@@ -91,6 +90,25 @@ case class ResolveDefaultColumns(catalog: SessionCatalog) extends Rule[LogicalPl
...
+   * list, with possible other unary operators like sorting and/or alias(es) in between.
+   */
+  private def insertsFromProject(i: InsertIntoStatement): Option[Project] = {
+    var node = i.query
+    def matches(node: LogicalPlan): Boolean = node match {
+      case _: GlobalLimit | _: LocalLimit | _: Offset | _: SubqueryAlias | _: Sort => true
+      case _ => false
+    }
+    while (matches(node)) {
+      node = node.children.head
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala:104]: Why getting the first child only? What if there is a join and one of the children has a project containing default?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala [hunk 2]
```diff
@@ -91,6 +90,25 @@ case class ResolveDefaultColumns(catalog: SessionCatalog) extends Rule[LogicalPl
...
   }
 
+  /**
+   * Checks if a logical plan is an INSERT INTO command where the inserted data comes from a SELECT
+   * list, with possible other unary operators like sorting and/or alias(es) in between.
+   */
+  private def insertsFromProject(i: InsertIntoStatement): Option[Project] = {
+    var node = i.query
+    def matches(node: LogicalPlan): Boolean = node match {
+      case _: GlobalLimit | _: LocalLimit | _: Offset | _: SubqueryAlias | _: Sort => true
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala:100]: Shall we resolve all the `Project`s?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala (hunk 1 above)
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala:104]: Good question, I added a test for this:

```
insert into t2 (i, s) select default, default from t1 inner join t1 using (i, s);

> true, 42L, "abc",
> true, 42L, "abc"
```

We want the `default` resolution to cover the topmost `select` list before the join, but not any below, per the specification. For example, this should not work:

```
insert into t2 (i, s) select i, s from (
  select default as i, default as s from t1)
  inner join t1
  using (i, s);
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala (hunk 2 above)
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala:100]: As I understand it, the intention is to allow and resolve `default` references in the top-most `Project`, but not such references within any table subqueries. For example, this should not work:

```
insert into t2 (i, s) select i, s from (
  select default as i, default as s from t1 order by i limit 1)
```

**Review by gengliangwang — APPROVED** [review]

**Comment by gengliangwang** [comment]
Thanks, merging to master/3.4

**Pull request closed without merging**